File: rpi/udp_handler.py
"""
udp_handler.py — Port of udp_handler.h

Listens on UDP_PORT for JSON commands from Q-SYS (or any UDP sender).
Runs in its own daemon thread so it never blocks the render loop.

Supported JSON commands (identical protocol to the ESP32 firmware):
  {"cmd":"text",       "seg":0, "text":"Hello", "color":"FFFFFF",
   "bgcolor":"000000", "font":"arial", "size":"auto",
   "align":"C", "effect":"none", "intensity":255}
  {"cmd":"layout",     "preset":1}          # 1-7, see config.LAYOUT_PRESETS
  {"cmd":"clear",      "seg":0}
  {"cmd":"clear_all"}
  {"cmd":"brightness", "value":200}
  {"cmd":"config",     "seg":0, "x":0, "y":0, "w":64, "h":32}
"""

import socket
import json
import threading
import logging
import os

from config import UDP_PORT, UDP_BIND_ADDR, MAX_TEXT_LENGTH, LAYOUT_PRESETS, LAYOUT_PRESETS_PORTRAIT, MAX_SEGMENTS, CONFIG_FILE, ORIENTATION, GROUP_ID
from segment_manager import SegmentManager

logger = logging.getLogger(__name__)

# Brightness is 0-255 in the protocol (same as ESP32 firmware).
# We convert to 0-100 for the rpi-rgb-led-matrix library when applying it.
_brightness = 128  # default
_brightness_lock = threading.Lock()

# Orientation is "landscape" or "portrait"
_orientation = ORIENTATION
_orientation_lock = threading.Lock()

# Group ID (0 = no grouping, 1-8 = assigned group)
_group_id = GROUP_ID
_group_id_lock = threading.Lock()


def get_brightness() -> int:
    with _brightness_lock:
        return _brightness


def _set_brightness(value: int):
    global _brightness
    with _brightness_lock:
        _brightness = max(0, min(255, int(value)))


def get_orientation() -> str:
    """Get current orientation: 'landscape' or 'portrait'"""
    with _orientation_lock:
        return _orientation


def get_canvas_dimensions() -> tuple[int, int]:
    """Get current canvas width and height based on orientation.
    Returns (width, height) for the virtual canvas.
    """
    from config import MATRIX_WIDTH, MATRIX_HEIGHT
    orientation = get_orientation()
    if orientation == "portrait":
        # Portrait: 32 wide × 64 tall
        return (MATRIX_HEIGHT, MATRIX_WIDTH)
    else:
        # Landscape: 64 wide × 32 tall
        return (MATRIX_WIDTH, MATRIX_HEIGHT)


def set_orientation(value: str):
    """Set orientation and persist to config file"""
    global _orientation
    value = value.lower()
    if value not in ("landscape", "portrait"):
        logger.warning(f"[ORIENTATION] Invalid value: {value}")
        return False
    
    with _orientation_lock:
        _orientation = value
    
    # Persist to config file
    _save_config()
    logger.info(f"[ORIENTATION] Set to: {value}")
    return True


def get_group_id() -> int:
    """Get current group ID (0 = no grouping, 1-8 = assigned group)"""
    with _group_id_lock:
        return _group_id


def set_group_id(value: int):
    """Set group ID and persist to config file"""
    global _group_id
    if not (0 <= value <= 8):
        logger.warning(f"[GROUP] Invalid group ID: {value}")
        return False
    
    with _group_id_lock:
        _group_id = value
    
    # Persist to config file
    _save_config()
    logger.info(f"[GROUP] Set to: {value}")
    return True


def _save_config():
    """Save current orientation and group ID to JSON config file"""
    try:
        os.makedirs(os.path.dirname(CONFIG_FILE), exist_ok=True)
        config_data = {
            "orientation": get_orientation(),
            "group_id": get_group_id()
        }
        with open(CONFIG_FILE, "w") as f:
            json.dump(config_data, f)
        logger.debug(f"[CONFIG] Saved to {CONFIG_FILE}")
    except Exception as e:
        logger.error(f"[CONFIG] Failed to save: {e}")


def _load_config():
    """Load orientation and group ID from JSON config file"""
    global _orientation, _group_id
    try:
        if os.path.exists(CONFIG_FILE):
            with open(CONFIG_FILE, "r") as f:
                config_data = json.load(f)
                orientation = config_data.get("orientation", "landscape")
                group_id = config_data.get("group_id", 0)
                with _orientation_lock:
                    _orientation = orientation
                with _group_id_lock:
                    _group_id = group_id
                logger.info(f"[CONFIG] Loaded orientation: {orientation}, group_id: {group_id}")
    except Exception as e:
        logger.warning(f"[CONFIG] Failed to load: {e}, using defaults")


class UDPHandler:
    """
    Receives JSON UDP packets and updates the SegmentManager.

    Call start() once — it spawns a background daemon thread.
    Call dispatch(raw_json_str) directly (e.g. from the web handler).
    """

    def __init__(self, segment_manager: SegmentManager,
                 brightness_callback=None,
                 orientation_callback=None):
        """
        :param segment_manager: shared SegmentManager instance
        :param brightness_callback: optional callable(int 0-255) invoked when
               brightness changes so the display can be updated immediately.
        :param orientation_callback: optional callable(str) invoked when
               orientation changes ('landscape' or 'portrait').
        """
        self._sm   = segment_manager
        self._sock = None
        self._running = False
        self._thread  = None
        self._brightness_callback = brightness_callback
        self._orientation_callback = orientation_callback
        self._orientation = get_orientation()  # Track current orientation for layout presets
        self._current_layout = 1  # Track current layout preset number (applies to both orientations)
        # Set to True on the first successfully dispatched command so
        # main.py can dismiss the IP splash screen.
        self._first_command_received = False

    # ─── Public API ────────────────────────────────────────────────────────

    def start(self):
        """Bind the UDP socket and start the listener thread."""
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._sock.bind((UDP_BIND_ADDR, UDP_PORT))
        self._sock.settimeout(1.0)   # unblocks the thread for clean shutdown
        self._running = True
        self._thread  = threading.Thread(target=self._run,
                                         name="udp-listener",
                                         daemon=True)
        self._thread.start()
        logger.info(f"[UDP] Listening on {UDP_BIND_ADDR}:{UDP_PORT}")

    def stop(self):
        self._running = False
        if self._sock:
            self._sock.close()

    def has_received_command(self) -> bool:
        """True once the first command has been dispatched (UDP or web)."""
        return self._first_command_received
    
    def get_current_layout(self) -> int:
        """Return the current layout preset number."""
        return self._current_layout

    def dispatch(self, raw: str):
        """Parse a raw JSON string and apply the command (thread-safe)."""
        try:
            doc = json.loads(raw)
        except json.JSONDecodeError as e:
            logger.error(f"[UDP] JSON parse error: {e}")
            return

        # Flag raised on the very first successfully parsed command —
        # used by main.py to dismiss the IP address splash screen.
        self._first_command_received = True

        # Check group filtering: if the command has a "group" field,
        # only execute if it matches this panel's group_id (or if group_id is 0)
        cmd_group = doc.get("group", 0)  # 0 = broadcast to all
        my_group = get_group_id()
        
        if cmd_group != 0 and my_group != 0 and cmd_group != my_group:
            # Command is for a different group, ignore it
            logger.debug(f"[UDP] Ignoring command for group {cmd_group} (this panel is group {my_group})")
            return

        cmd = doc.get("cmd", "")

        if cmd == "text":
            seg     = int(doc.get("seg", 0))
            text    = str(doc.get("text", ""))[:MAX_TEXT_LENGTH]
            color   = str(doc.get("color",   "FFFFFF"))
            bgcolor = str(doc.get("bgcolor", "000000"))
            align   = str(doc.get("align",   "C"))
            effect  = str(doc.get("effect",  "none"))
            intens  = int(doc.get("intensity", 255))
            self._sm.update_text(seg, text, color=color, bgcolor=bgcolor,
                                 align=align, effect=effect, intensity=intens)

        elif cmd == "layout":
            preset = int(doc.get("preset", 1))
            self._apply_layout(preset)

        elif cmd == "clear":
            seg = int(doc.get("seg", 0))
            self._sm.clear_segment(seg)

        elif cmd == "clear_all":
            self._sm.clear_all()

        elif cmd == "brightness":
            val = doc.get("value", -1)
            if 0 <= int(val) <= 255:
                _set_brightness(int(val))
                if self._brightness_callback:
                    self._brightness_callback(int(val))

        elif cmd == "orientation":
            value = str(doc.get("value", "landscape"))
            if set_orientation(value):
                self._orientation = value  # Update instance variable for layout presets
                if self._orientation_callback:
                    self._orientation_callback(value)

        elif cmd == "group":
            # Set this panel's group ID
            value = int(doc.get("value", 0))
            set_group_id(value)

        elif cmd == "config":
            seg = int(doc.get("seg", 0))
            x   = int(doc.get("x", 0))
            y   = int(doc.get("y", 0))
            w   = int(doc.get("w", 64))
            h   = int(doc.get("h", 32))
            self._sm.configure(seg, x, y, w, h)

        else:
            logger.warning(f"[UDP] Unknown cmd: {cmd}")

    # ─── Layout presets ────────────────────────────────────────────────────

    def _apply_layout(self, preset: int):
        """Apply a numbered layout preset from config.LAYOUT_PRESETS or LAYOUT_PRESETS_PORTRAIT."""
        # Use portrait layouts if in portrait mode
        if self._orientation == "portrait":
            zones = LAYOUT_PRESETS_PORTRAIT.get(preset)
        else:
            zones = LAYOUT_PRESETS.get(preset)
        
        if zones is None:
            logger.warning(f"[UDP] Unknown layout preset {preset}")
            return
        
        # Remember the current layout preset number
        self._current_layout = preset
        
        logger.info(f"[UDP] LAYOUT preset={preset} ({len(zones)} segment(s)) in {self._orientation} mode")
        for i in range(MAX_SEGMENTS):
            if i < len(zones):
                x, y, w, h = zones[i]
                self._sm.configure(i, x, y, w, h)
                self._sm.activate(i, True)
            else:
                self._sm.activate(i, False)

    # ─── Internal ──────────────────────────────────────────────────────────

    def _run(self):
        while self._running:
            try:
                data, addr = self._sock.recvfrom(4096)
                raw = data.decode("utf-8", errors="replace").strip()
                logger.debug(f"[UDP] {addr[0]}:{addr[1]}  {raw[:120]}")
                self.dispatch(raw)
            except socket.timeout:
                continue
            except OSError:
                break   # socket closed by stop()
        logger.info("[UDP] Listener thread exited")

File: rpiC++/udp_handler.cpp
// udp_handler.cpp - UDP JSON command parser implementation

#include "udp_handler.h"
#include "config.h"
#include <nlohmann/json.hpp>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <cstring>
#include <fstream>
#include <iostream>
#include <sys/stat.h>
#include <libgen.h>

using json = nlohmann::json;

UDPHandler::UDPHandler(SegmentManager* segment_manager,
                       BrightnessCallback brightness_cb,
                       OrientationCallback orientation_cb)
    : sm_(segment_manager),
      socket_fd_(-1),
      running_(false),
      first_command_received_(false),
      brightness_callback_(brightness_cb),
      orientation_callback_(orientation_cb),
      orientation_(LANDSCAPE),
      current_layout_(1),
      brightness_(128),
      group_id_(0) {
    loadConfig();
}

UDPHandler::~UDPHandler() {
    stop();
}

void UDPHandler::start() {
    socket_fd_ = socket(AF_INET, SOCK_DGRAM, 0);
    if (socket_fd_ < 0) {
        std::cerr << "[UDP] Failed to create socket" << std::endl;
        return;
    }
    
    int reuse = 1;
    setsockopt(socket_fd_, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse));
    
    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(UDP_PORT);
    addr.sin_addr.s_addr = INADDR_ANY;
    
    if (bind(socket_fd_, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        std::cerr << "[UDP] Failed to bind to port " << UDP_PORT << std::endl;
        close(socket_fd_);
        socket_fd_ = -1;
        return;
    }
    
    // Set receive timeout for clean shutdown
    struct timeval tv;
    tv.tv_sec = 1;
    tv.tv_usec = 0;
    setsockopt(socket_fd_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
    
    running_ = true;
    listener_thread_ = std::thread(&UDPHandler::run, this);
    
    std::cout << "[UDP] Listening on " << UDP_BIND_ADDR << ":" << UDP_PORT << std::endl;
}

void UDPHandler::stop() {
    running_ = false;
    if (socket_fd_ >= 0) {
        close(socket_fd_);
        socket_fd_ = -1;
    }
    if (listener_thread_.joinable()) {
        listener_thread_.join();
    }
}

void UDPHandler::run() {
    char buffer[4096];
    
    while (running_) {
        struct sockaddr_in client_addr;
        socklen_t client_len = sizeof(client_addr);
        
        ssize_t len = recvfrom(socket_fd_, buffer, sizeof(buffer) - 1, 0,
                              (struct sockaddr*)&client_addr, &client_len);
        
        if (len < 0) {
            if (!running_) break;
            continue;
        }
        
        buffer[len] = '\0';
        std::string raw(buffer, len);
        
        // Trim whitespace
        size_t start = raw.find_first_not_of(" \t\r\n");
        size_t end = raw.find_last_not_of(" \t\r\n");
        if (start != std::string::npos && end != std::string::npos) {
            raw = raw.substr(start, end - start + 1);
        }
        
        dispatch(raw);
    }
    
    std::cout << "[UDP] Listener thread exited" << std::endl;
}

void UDPHandler::dispatch(const std::string& raw_json) {
    std::cout << "[UDP] Received: " << raw_json << std::endl;
    
    try {
        json doc = json::parse(raw_json);
        first_command_received_ = true;
        
        std::string cmd = doc.value("cmd", "");
        
        // Check group filtering
        int cmd_group = doc.value("group", 0);
        int my_group = group_id_;
        
        if (cmd_group != 0 && my_group != 0 && cmd_group != my_group) {
            std::cout << "[UDP] Ignoring command for group " << cmd_group 
                     << " (this panel is group " << my_group << ")" << std::endl;
            return;
        }
        
        std::cout << "[UDP] Executing command: " << cmd 
                 << " (group=" << cmd_group << ", my_group=" << my_group << ")" << std::endl;
        
        if (cmd == "text") {
            int seg = doc.value("seg", 0);
            std::string text = doc.value("text", "");
            std::string color = doc.value("color", "FFFFFF");
            std::string bgcolor = doc.value("bgcolor", "000000");
            std::string align = doc.value("align", "C");
            std::string effect = doc.value("effect", "none");
            int intensity = doc.value("intensity", 255);
            std::string font = doc.value("font", "arial");  // "arial" or "monospace"
            
            sm_->updateText(seg, text, color, bgcolor, align, effect, intensity, font);
            
        } else if (cmd == "layout") {
            int preset = doc.value("preset", 1);
            applyLayout(preset);
            
        } else if (cmd == "clear") {
            int seg = doc.value("seg", 0);
            sm_->clearSegment(seg);
            
        } else if (cmd == "clear_all") {
            sm_->clearAll();
            
        } else if (cmd == "brightness") {
            int val = doc.value("value", -1);
            if (val >= 0 && val <= 255) {
                std::lock_guard<std::mutex> lock(config_mutex_);
                brightness_ = val;
                if (brightness_callback_) {
                    brightness_callback_(val);
                }
                sm_->markAllDirty();
                saveConfig();
            }
            
        } else if (cmd == "orientation") {
            std::string value = doc.value("value", "landscape");
            std::transform(value.begin(), value.end(), value.begin(), ::tolower);
            
            Orientation new_orient = (value == "portrait") ? PORTRAIT : LANDSCAPE;
            {
                std::lock_guard<std::mutex> lock(config_mutex_);
                orientation_ = new_orient;
            }
            
            if (orientation_callback_) {
                orientation_callback_(new_orient);
            }
            
            // Reapply current layout for new orientation
            applyLayout(current_layout_);
            saveConfig();
            
        } else if (cmd == "group") {
            int value = doc.value("value", 0);
            if (value >= 0 && value <= 8) {
                std::lock_guard<std::mutex> lock(config_mutex_);
                group_id_ = value;
                sm_->markAllDirty();
                saveConfig();
            }
            
        } else if (cmd == "config") {
            int seg = doc.value("seg", 0);
            int x = doc.value("x", 0);
            int y = doc.value("y", 0);
            int w = doc.value("w", 64);
            int h = doc.value("h", 32);
            sm_->configure(seg, x, y, w, h);
            
        } else if (cmd == "frame") {
            int seg = doc.value("seg", 0);
            bool enabled = doc.value("enabled", false);
            std::string color = doc.value("color", "FFFFFF");
            int width = doc.value("width", 2);
            sm_->setFrame(seg, enabled, color, width);
            
        } else {
            std::cerr << "[UDP] Unknown cmd: " << cmd << std::endl;
        }
        
    } catch (const json::exception& e) {
        std::cerr << "[UDP] JSON parse error: " << e.what() << std::endl;
    }
}

void UDPHandler::applyLayout(int preset) {
    if (preset < 1 || preset > 14) {
        std::cerr << "[UDP] Unknown layout preset " << preset << std::endl;
        return;
    }
    
    // Skip if layout didn't actually change
    if (current_layout_ == preset) {
        return;  // No-op, already on this layout
    }
    
    current_layout_ = preset;
    
    const std::vector<LayoutRect>* zones;
    if (orientation_ == PORTRAIT) {
        zones = &LAYOUT_PORTRAIT[preset];
    } else {
        zones = &LAYOUT_LANDSCAPE[preset];
    }
    
    std::cout << "[UDP] LAYOUT preset=" << preset 
             << " (" << zones->size() << " segment(s)) in "
             << (orientation_ == PORTRAIT ? "portrait" : "landscape") << " mode" << std::endl;
    
    for (int i = 0; i < MAX_SEGMENTS; i++) {
        if (i < (int)zones->size()) {
            const LayoutRect& rect = (*zones)[i];
            sm_->configure(i, rect.x, rect.y, rect.w, rect.h);
            sm_->activate(i, true);
        } else {
            sm_->activate(i, false);
        }
    }
}

void UDPHandler::loadConfig() {
    std::ifstream file(CONFIG_FILE);
    if (!file.is_open()) {
        std::cout << "[CONFIG] No config file found, using defaults" << std::endl;
        return;
    }
    
    try {
        json config;
        file >> config;
        
        std::string orient = config.value("orientation", "landscape");
        orientation_ = (orient == "portrait") ? PORTRAIT : LANDSCAPE;
        
        group_id_ = config.value("group_id", 0);
        brightness_ = config.value("brightness", 128);
        
        std::cout << "[CONFIG] Loaded orientation: " << orient 
                 << ", group_id: " << group_id_ 
                 << ", brightness: " << brightness_ << std::endl;
    } catch (const json::exception& e) {
        std::cerr << "[CONFIG] Failed to parse: " << e.what() << std::endl;
    }
}

void UDPHandler::saveConfig() {
    // Create directory if it doesn't exist
    char* path_copy = strdup(CONFIG_FILE);
    char* dir = dirname(path_copy);
    mkdir(dir, 0755);
    free(path_copy);
    
    json config;
    {
        std::lock_guard<std::mutex> lock(config_mutex_);
        config["orientation"] = (orientation_ == PORTRAIT) ? "portrait" : "landscape";
        config["group_id"] = group_id_;
        config["brightness"] = brightness_;
    }
    
    std::ofstream file(CONFIG_FILE);
    if (file.is_open()) {
        file << config.dump(2);
        std::cout << "[CONFIG] Saved to " << CONFIG_FILE << std::endl;
    } else {
        std::cerr << "[CONFIG] Failed to save" << std::endl;
    }
}
